File: exercise/39/solution.cpp
class Solution {
public:
    void dfs(vector<int> &candidates, int index, int res, vector<vector<int>> &result, vector<int> &now) {
        if (res == 0) {
            result.push_back(now);
            return;
        }
        if (candidates[index] > res) {
            return;
        } else {
            now.push_back(candidates[index]);
            dfs(candidates, index, res - candidates[index], result, now);
            now.pop_back();
            if (index < candidates.size() - 1)
                dfs(candidates, index + 1, res, result, now);
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> now;
        dfs(candidates, 0, target, result, now);
        return result;
    }
};

File: exercise/39/solution.py
class Solution:
    def combinationSumPart(self, candidates, target, index, result, now):
        if target == 0:
            result.append(now)
        else:
            if index == len(candidates) or candidates[index] > target:
                return
            now.append(candidates[index])
            self.combinationSumPart(candidates, target - candidates[index], index + 1, result, now)
            now.pop()
            next_index = index + 1
            while next_index < len(candidates) and candidates[next_index] == candidates[next_index-1]:
                next_index += 1
            self.combinationSumPart(candidates, target, next_index, result, now)

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        now = []
        self.combinationSumPart(candidates, target, 0, result, now)
        return result

